File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include<iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    
    int tyear = 0, tmonth = 0, tday = 0;
    for(int j = 0; j < 4; j++){
        tyear *= 10;
        tyear += today[j] - '0';
    }
    for(int j = 5; j < 7; j++){
        tmonth *= 10;
        tmonth += today[j] - '0';
    }
    for(int j= 8; j < 10; j++){
        tday *= 10;
        tday += today[j] - '0';
    }
    
    int ttday = 0;
    ttday += tyear * 12 * 28 + tmonth * 28 + tday;
    
    
    
    for(int i = 0; i < privacies.size(); i++){
        int year = 0, month = 0, day = 0;
        for(int j = 0; j < 4; j++){
            year *= 10;
            year += privacies[i][j] - '0';
        }
        for(int j = 5; j < 7; j++){
            month *= 10;
            month += privacies[i][j] - '0';
        }
        for(int j= 8; j < 10; j++){
            day *= 10;
            day += privacies[i][j] - '0';
        }
        
        int tday =  year * 12 * 28 + month * 28 + day;
        
        char now = privacies[i][11];
        
        for(int j = 0; j < terms.size(); j++){
            if(terms[j][0] == now){
                int m = 0;
                for(int k = 2; k < terms[j].size(); k++){
                    m *= 10;
                    m += terms[j][k] - '0';
                }
                tday += m * 28;
                break;
            }
        }
        
        if(tday - 1 < ttday){
            answer.push_back(i+1);
        }
    }    
    
    return answer;
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {
    
    public boolean expire_day(String today, String day, int term) {
        String[] tod = today.split("\\.");
        String[] d = day.split("\\.");
        
        int ty = Integer.parseInt(tod[0]);
        int tm = Integer.parseInt(tod[1]);
        int td = Integer.parseInt(tod[2]);
        int ny = Integer.parseInt(d[0]);
        int nm = Integer.parseInt(d[1]);
        int nd = Integer.parseInt(d[2]);
        
        nm += term;
        while(nm > 12) {
            nm -= 12;
            ny++;
        }
        
        if(nd == 1) {
            if((nm <= 7 && nm % 2 == 1) || (nm >= 8 && nm % 2 == 0)) nd = 31;
            else nd = 30;
            
            if(--nm == 0) {
                nm = 12;
                ny--;
            }
        } 
        else {
            nd--;
        }
        
        // today가 더 크면 true == 파괴
        if(ty > ny) return true;
        else if(ty == ny) {
            if(tm > nm) return true;
            else if(tm == nm) {
                if(td > nd) return true;
            }
        }
        
        return false;
    }
    
    public int[] solution(String today, String[] terms, String[] privacies) {
        int[] answer = {};
        
        List<Integer> res = new ArrayList<>();
        Map<String, Integer> m = new HashMap<>();
        
        for(String str : terms) {
            String[] s = str.split(" ");
            m.put(s[0], Integer.parseInt(s[1]));
        }
        
        for(int i = 0; i < privacies.length; ++i) {
            String[] s = privacies[i].split(" ");
            
            if(expire_day(today, s[0], m.get(s[1]))) {
                res.add(i + 1);
            }
        }
        
        answer = res.stream().mapToInt(i -> i).toArray();
        
        return answer;
    }
}